site/app/controllers/forum/ForumController.php

```diff
@@ -369,6 +369,38 @@ public function publishThread() {
         return $this->core->getOutput()->renderJsonSuccess($result);
     }
 
+    /**
+     * @Route("/courses/{_semester}/{_course}/forum/make_announcement", methods={"POST"})
+     * @AccessControl(permission="forum.modify_announcement")
+     */
+    public function makeAnnouncement() {
+        if (!isset($_POST['id'])) {
+            return;
+        }
+        // Check that the post is the first post of the thread
+        $thread_info = $this->core->getQueries()->findPost($_POST['id']);
+        if ($thread_info["parent_id"] != -1) {
+            return;
+        }
+        // Check that the post is indeed less than an hour old on the server
+        $dateTime = new \DateTime($thread_info['timestamp']);
+        $now = $this->core->getDateTimeNow();
+
+        if ($dateTime->add(new \DateInterval("PT1H")) < $now) {
+            return;
+        }
+
+        $full_course_name = $this->core->getFullCourseName();
+        $thread_title = trim($_POST["title"]);
+        $thread_post_content = str_replace("\r", "", $_POST["thread_post_content"]);
+        $metadata = json_encode(['url' => $this->core->buildCourseUrl(['forum', 'threads', $_POST['id']]), 'thread_id' => $_POST['id']]);
+
+        $subject = "New Announcement: " . Notification::textShortner($thread_title);
+        $content = "An Instructor or Teaching Assistant made an announcement in:\n" . $full_course_name . "\n\n" . $thread_title . "\n\n" . $thread_post_content;
+        $event = ['component' => 'forum', 'metadata' => $metadata, 'content' => $content, 'subject' => $subject];
+        $this->core->getNotificationFactory()->onNewAnnouncement($event);
+    }
```

Comment:
This should return a json response or something. You have it returning void while not providing any information to the user of what went wrong. You should also have some sort of success message.

Reply:
Should it simply return a json or should it display a message at the top too? Right now, the only indication that it succeeded is if the user checks to see if the button still exists, or if they go into the email status page if they are an instructor.

Reply:
Since the site is making a direct request here a display message at the top would probably be better.